File: 프로그래머스/3/12927. 야근 지수/야근 지수.cpp
#include <string>
#include <vector>
#include <queue>
using namespace std;

long long solution(int n, vector<int> works) {
    long long answer = 0;
    priority_queue<int> pq;
    for(int i=0;i<works.size();i++){
        pq.push(works[i]);
    }
    while(!pq.empty()&&n){
        int a=pq.top();
        pq.pop();
        if(a>1){
            pq.push(a-1);
        }
        n--;
    }
    while(!pq.empty()){
        long long a=pq.top();
        pq.pop();
        answer+=a*a;
    }
    return answer;
}

File: 프로그래머스/3/12927. 야근 지수/야근 지수.java
import java.util.*;

class Solution {
    public long solution(int n, int[] works) {
        long answer = 0;
        Queue<Integer> pq = new PriorityQueue<>((n1, n2) -> n1 - n2);
        for (int i = 0; i < works.length; i++) {
            pq.add(works[i]);
        }
        while (!pq.isEmpty() && n > 0) {
            int a = pq.poll();
            if (a > 1) {
                pq.add(a - 1);
            }
            n--;
        }
        while (!pq.isEmpty()) {
            long a = pq.poll();
            answer += a * a;
        }
        return answer;
    }
}